codalab/lib/canonicalize.py

```diff
@@ -115,12 +115,7 @@ def get_target_path(bundle_store, model, target):
     Return the on-disk location of the target (bundle_uuid, subpath) pair.
     """
     (uuid, path) = target
-    bundle = model.get_bundle(uuid)
-    if not bundle.data_hash:
-        # Note that the bundle might not be ready, but return the location anyway to the temporary directory.
-        bundle_root = get_current_location(bundle_store, uuid)
```

Comment:
Thinking about this some more:

When the job is being run, there are other files that created too in temp: <uuid>.status, etc.  Do we want these hanging out in the bundle store?  The logic is certainly simpler that way, and when you rm a bundle, it could delete <uuid>.* rather than currently having a slew of temp files just hanging out there.  (Some of these shouldn't actually be deleted because their states are still 'running' but their workers died, so their data is stuck in temp.)

On the other hand, the bundle store would be constantly changing.  I think this is okay though, because I think the contents of a bundle (being run or being uploaded) should be accessible as is, and we should just use the data_hash and state flag to denote whether a bundle's contents have actually finished changing and become immutable.  We can even do fancy things like partial uploads or depend on a running bundle later.

So in short, let's just use the data directory, not the temp directory everywhere (need to change this in upload logic too).  In general, copying things to a temp directory first and then moving atomically is good (how browser downloads are handled), but this is because there's no natural bit that tells you whether a file is complete, but here we rely on the DB (and should make sure that logic is sound).

Do you see any problems with this?